src/cargo/ops/cargo_package.rs

```diff
@@ -186,34 +256,102 @@ pub fn package(ws: &Workspace<'_>, opts: &PackageOpts<'_>) -> CargoResult<Option
             }
         } else {
-            let tarball = create_package(ws, pkg, ar_files)?;
-            if opts.verify {
-                run_verify(ws, pkg, &tarball, &opts)
-                    .with_context(|| "failed to verify package tarball")?;
-            }
-            dsts.push(tarball);
+            let tarball = create_package(ws, &pkg, ar_files, Some(&local_reg))?;
+            local_reg.add_package(&pkg, &tarball)?;
+            outputs.push((pkg, opts, tarball));
         }
     }
 
-    if opts.list {
-        // We're just listing, so there's no file output
-        Ok(None)
-    } else {
-        Ok(Some(dsts))
+    // Verify all packages in the workspace. This can be done in any order, since the dependencies
+    // are already all in the local registry overlay.
+    if opts.verify {
+        for (pkg, opts, tarball) in &outputs {
+            run_verify(ws, pkg, tarball, Some(&local_reg), opts)
+                .with_context(|| "failed to verify package tarball")?
+        }
     }
+
+    Ok(outputs.into_iter().map(|x| x.2).collect())
+}
+
+/// Just the part of the dependency graph that's between the packages we're packaging.
+/// (Is the package name a good key? Does it uniquely identify packages?)
+#[derive(Clone, Debug, Default)]
+struct LocalDependencies {
+    packages: HashMap<PackageId, (Package, CliFeatures)>,
+    graph: Graph<PackageId, ()>,
+}
+
+impl LocalDependencies {
+    fn sort(&self) -> Vec<(Package, CliFeatures)> {
+        self.graph
+            .sort()
+            .into_iter()
+            .map(|name| self.packages[&name].clone())
+            .collect()
+    }
+}
+
+/// Build just the part of the dependency graph that's between the given packages,
+/// ignoring dev dependencies.
+///
+/// We assume that the packages all belong to this workspace.
+fn local_deps(packages: impl Iterator<Item = (Package, CliFeatures)>) -> LocalDependencies {
+    let packages: HashMap<PackageId, (Package, CliFeatures)> =
+        packages.map(|pkg| (pkg.0.package_id(), pkg)).collect();
+
+    // Dependencies have source ids but not package ids. We draw an edge
+    // whenever a dependency's source id matches one of our packages. This is
+    // wrong in general because it doesn't require (e.g.) versions to match. But
+    // since we're working only with path dependencies here, it should be fine.
+    let source_to_pkg: HashMap<_, _> = packages
+        .keys()
+        .map(|pkg_id| (pkg_id.source_id(), *pkg_id))
+        .collect();
+
+    let mut graph = Graph::new();
+    for (pkg, _features) in packages.values() {
+        graph.add(pkg.package_id());
+        for dep in pkg.dependencies() {
+            // Ignore dev-dependencies because they aren't needed for intra-workspace
+            // lockfile generation or verification.
+            if dep.kind() == DepKind::Development || !dep.source_id().is_path() {
+                continue;
+            };
+
```

Comment:
```suggestion
            // Ignore local dev-dependencies because they aren't needed for intra-workspace
            // lockfile generation or verification as they get stripped on publish
            if dep.kind() == DepKind::Development || !dep.source_id().is_path() {
                continue;
            };

```